app/src/main/java/com/github/sdp/tarjetakuna/database/UserRTDB.kt

```diff
@@ -0,0 +1,148 @@
+
+    init {
+        this.db = database.usersTable()
+        cardsRTDB = CardsRTDB(database)
+    }
+
+    /**
+     * Adds a card to the users collection.
+     */
+    fun addCard(card: DBMagicCard, userUID: String) {
+        val cardUID = card.code + card.number
+        val fbpossession = card.possession.toString().lowercase()
+        db.child(userUID).child(fbpossession).child(cardUID).setValue(1)
+        cardsRTDB.addCardToGlobalCollection(card)
+    }
+
+    /**
+     * Adds a list of cards to the users collection.
+     */
+    fun addMultipleCards(cards: List<DBMagicCard>, userUID: String, possession: CardPossession) {
+        for (card in cards) {
+            addCard(card, userUID)
+        }
+    }
+
+    /**
+     * Removes a card from the users collection.
+     */ //todo: handle multiple cards and how this interacts with global collection
+    fun removeCard(userUID: String, card: DBMagicCard) {
+        val fbPosession = card.possession.toString().lowercase()
+        val cardUID = card.code + card.number
+        db.child(userUID).child(fbPosession).child(cardUID)
+            .removeValue()
+    }
+
+    /**
+     * Get the unique card code from the user's collection asynchronously from the database (based on possession category)
+     */
+    private fun getCardCodeFromUserCollection(
+        userUID: String,
+        cardUID: String,
+        possession: CardPossession
+    ): CompletableFuture<DataSnapshot> {
+        val future = CompletableFuture<DataSnapshot>()
+        db.child(userUID).child(possession.toString().lowercase()).child(cardUID).get()
+            .addOnSuccessListener {
+                if (it.value == null) {
+                    future.completeExceptionally(NoSuchFieldException("card $cardUID is not in user collection"))
+                } else {
+                    future.complete(it)
+                }
+            }.addOnFailureListener {
+                future.completeExceptionally(it)
+            }
+        return future
+    }
+
+    /**
+     * Get the card (json) from the user's collection asynchronously from the database (based on possession category)
+     */
+    //@RequiresApi(Build.VERSION_CODES.S)//for exceptionally
+    fun getCardFromUserPossession(
+        userUID: String,
+        setCode: String,
+        setNumber: Int,
+        possession: CardPossession
+    ): CompletableFuture<DataSnapshot> {
+        val cardUID = setCode + setNumber
+        var future = CompletableFuture<DataSnapshot>()
+        getCardCodeFromUserCollection(userUID, cardUID, possession).thenAccept {
+            future =
+                cardsRTDB.getCardFromGlobalCollection(cardUID) //only get the card if the user has it in their collection
+        }//todo: not sure how to handle exception case (needs certain api level), is it ok to just not handle this ?
+//            .exceptionally { exp = CompletableFuture.failedFuture<DataSnapshot>(it.cause)
+//            null}

```

Comment:
I'm not sure, but couldn't you handle the exception case like this?

`        getCardCodeFromUserCollection(userUID, cardUID, possession).thenAccept {
            future =
                cardsRTDB.getCardFromGlobalCollection(cardUID) //only get the card if the user has it in their collection
        }.exceptionally { e ->
            future.completeExceptionally(e)
            null 
        }  `